graphistry/plugins/spannergraph.py

```diff
@@ -0,0 +1,267 @@
+import os
+import pandas as pd
+import json
+import time
+from typing import Any, List, Dict, Optional
+
+from graphistry.Plottable import Plottable
+
+from graphistry.util import setup_logger
+logger = setup_logger(__name__)
+
+
+class SpannerConnectionError(Exception):
+    """Custom exception for errors related to Spanner connection."""
+    pass
+
+class SpannerQueryResult:
+    """
+    Encapsulates the results of a query, including metadata.
+
+    :ivar list data: The raw query results.
+    """
+
+    def __init__(self, data: List[Any], column_names: List[str]):
+        """
+        Initializes a SpannerQueryResult instance.
+
+        :param data: The raw query results.
+        :type List[Any]
+        :param column_names: a list of the column names from the cursor, defaults to None 
+        :type: List[str]
+        """
+        self.data = data
+        self.column_names = column_names
+
+
+class SpannerGraph:
+    """
+    A comprehensive interface for interacting with Google Spanner Graph databases.
+
+    :ivar str project_id: The Google Cloud project ID.
+    :ivar str instance_id: The Spanner instance ID.
+    :ivar str database_id: The Spanner database ID.
+    :ivar Any connection: The active connection to the Spanner database.
+    :ivar Any graphistry: The Graphistry parent object.
+    """
+
+    def __init__(self, g: Plottable, project_id: str, instance_id: str, database_id: str, credentials_file: Optional[str] = None):
+        """
+        Initializes the SpannerGraph instance.
+
+        :param graphistry: The Graphistry parent object.
+        :param project_id: The Google Cloud project ID.
+        :param instance_id: The Spanner instance ID.
+        :param database_id: The Spanner database ID.
+        """
+        self.g = g
```

Comment:
see comment https://github.com/graphistry/pygraphistry/pull/622/files#r1922867829

i'm tempted to:

- get rid of member field `g` here, as `SpannerGraph()` seems to persist across `g1, g2, g3,...`compositions, while this `g` would be stale

- instead thread through `g` through the callers using it

- an alternative to achieve this is to make this class a mixin, so `self: SpannerGraph <: Plottable`, but that's more complicated than we want rn imo